### PR TITLE
[NPUW][SWA][PR2] Attention mask generation and KV slice writes

### PR DESCRIPTION
**Pull request by GuoliangShiIntel**
### Details:
This PR adds the runtime utility layer for Hybrid Sliding Window Attention (SWA) in NPUW, introducing standalone helper functions for attention mask construction and KV cache slice writes.

**It provides:**
1. Mask generation: fill_causal_sliding_mask / overlay_vision_bidirectional_mask / fill_attention_masks build the causal + sliding-window (and optional vision bidirectional) attention masks consumed by SDPA.
2. KV slice writes: write_swa_kv_slice_left_aligned / write_swa_kv_slice_circular write a new KV slice into a left-aligned or circular sliding-window KV cache buffer.

**Scope of this PR:** Pure runtime utility functions with unit tests for mask generation and KV slice write correctness. Has no dependency on the `ShrinkSlidingWindowKVCache` transformation or SWA layout detection from the prior PR, and can be reviewed/merged independently.

Does not include pipeline/inference-request wiring yet (that is handled in a following PR).

image

### Tickets:
 - [*EISW-232937*](https://jira.devtools.intel.com/browse/EISW-232937)

### AI Assistance:
 - *AI assistance used: no / yes*
 - *If yes, summarize how AI was used and what human validation was performed (build/tests/manual checks).*

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+
+MaskView get_mask_view(const ov::SoPtr<ov::ITensor>& mask_tensor,
+                       uint32_t num_real_new_tokens,
+                       const char* caller_name) {
+    OPENVINO_ASSERT(mask_tensor->get_element_type() == ov::element::f32,
+                    caller_name,
+                    ": Attention mask tensor is expected to be f32, got: ",
+                    mask_tensor->get_element_type());
+    const auto& shape = mask_tensor->get_shape();
+    OPENVINO_ASSERT(shape.size() >= 2, caller_name, ": Attention mask tensor rank must be >= 2, got shape: ", shape);
```
[the remark this answers is not in the document]
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:39]: Fixed

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+                                          const std::unordered_map<std::string, ov::Output<const ov::Node>>& in_ports,
+                                          uint32_t num_stored_tokens_before,
+                                          uint32_t num_real_new_tokens,
+                                          uint32_t window_size,
+                                          const int64_t* token_type_ids_real) {
+    const auto it = in_ports.find(ov::npuw::util::kSlidingWindowAttentionMaskParamName);
+    if (it == in_ports.end()) {
+        return;
+    }
+    auto mask_tensor = request->get_tensor(it->second);
```
[the remark this answers is not in the document]
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:363]: without change

**Review by GuoliangShiIntel — COMMENTED** [review]
**self comment:**

Since the mask generation logic is relatively complex, would it make sense to add a debug helper to visualize the generated mask (e.g., rows/columns and layout)? This could help quickly validate whether the mask is constructed as expected and make troubleshooting easier.

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+                    ": num_real_new_tokens (",
+                    num_real_new_tokens,
+                    ") exceeds query axis (",
+                    row_dim,
+                    ")");
+
+    MaskView mv;
+    mv.row_dim = row_dim;
+    mv.col_dim = col_dim;
+    mv.past_width = col_dim - row_dim;
```
Comment (intelgaoxiong) [inline src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:69]: This should be incorrect when mask shape is [1, past + present]
Present is chunk_size, but Q is single token when last token slice hoisting happens (https://github.com/openvinotoolkit/openvino/pull/37609).

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+        auto dst_first_leg = ov::npuw::util::make_tensor_slice(dst_tensor, dst_kv_dim, dst_start, capacity);
+        ov::npuw::util::copy_tensor_by_dim(src_first_leg, dst_first_leg, src_kv_dim, dst_kv_dim);
+
+        auto src_second_leg = ov::npuw::util::make_tensor_slice(src_slice, src_kv_dim, first_leg_len, tokens_to_write);
+        auto dst_second_leg = ov::npuw::util::make_tensor_slice(dst_tensor, dst_kv_dim, 0u, second_leg_len);
+        ov::npuw::util::copy_tensor_by_dim(src_second_leg, dst_second_leg, src_kv_dim, dst_kv_dim);
+    }
+}
+
+void ov::npuw::util::write_swa_kv_slice_left_aligned(ov::SoPtr<ov::ITensor> dst_tensor,
```
Comment (intelgaoxiong) [inline src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:123]: Could you please explain the scenarios of using `write_swa_kv_slice_left_aligned` and `write_swa_kv_slice_circular`?

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+    const uint32_t src_len = static_cast<uint32_t>(src_new_kv->get_shape()[src_kv_dim]);
+    const uint32_t tokens_to_write = std::min({num_new_tokens, new_valid, src_len});
+
+    // keep: number of old tokens that remain visible after appending the new chunk.
+    // If keep < old_valid, the window is saturated and we must shift surviving old
+    // tokens to the front before appending new ones.
+    const uint32_t keep = new_valid - tokens_to_write;
+    const bool needs_shift = (keep > 0 && keep < old_valid);
+
+    if (needs_shift && dst_kv_dim == 3u) {
```
Comment (intelgaoxiong) [inline src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:149]: Can we use circle buffer as well to avoid shift?

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+                         ? ov::npuw::util::make_tensor_slice(src_new_kv, src_kv_dim, src_len - tokens_to_write, src_len)
+                         : src_new_kv;
+    auto dst_back = ov::npuw::util::make_tensor_slice(dst_tensor, dst_kv_dim, keep, keep + tokens_to_write);
+    ov::npuw::util::copy_tensor_by_dim(src_slice, dst_back, src_kv_dim, dst_kv_dim);
+}
+
+namespace {
+
+template <typename T>
+void fill_causal_sliding_window_mask_typed(const MaskView& mask_view,
```
Comment (intelgaoxiong) [inline src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:208]: Do you think it's more readable if split to 2 parts:
1. compute visible circle past slots:
```
struct SlotRange { uint32_t start; uint32_t end; };

std::vector<SlotRange> compute_visible_past_slots() {
    std::vector<SlotRange> ranges;
    if (past_width == 0) return ranges; // empty

    if (unsaturated) {
        // calc begin, end
        // single range
        return ranges;
    }

    //  saturated case:
     // calc begin, end
     // 2 ranges - wrapped

    return ranges;
}
```
2. populate masked and attended values:
```
2.1 init row: all masked
2.2 attend past in Ranges (from step 1)
2.3 attend present
```

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+    fill_causal_sliding_window_mask(mask_tensor, num_stored_tokens_before, num_real_new_tokens, window_size);
+    if (token_type_ids_real != nullptr) {
+        overlay_vision_bidirectional_mask(mask_tensor, token_type_ids_real, num_real_new_tokens);
+    }
+}
+
+namespace {
+
+template <typename T>
+void overlay_vision_bidirectional_mask_typed(const MaskView& mask_view,
```
Comment (intelgaoxiong) [inline src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:373]: When will this be used in pipeline?

Comment (intelgaoxiong) [reply src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:373]: I see, it's fill_attention_masks

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp (hunk 3 above)
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:69]: Valid comment. This change only applies to the Gemma4 E2B and E4B models, and neither of these models currently supports SWA with the HFA size blocker. In addition, there is no available condition to verify that this change is correct.

Therefore, I prefer to keep this as a required follow up task and address it when SWA is enabled for the E2B and E4B models.

JIRA: https://jira.devtools.intel.com/browse/EISW-232940

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp (hunk 4 above)
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:123]: `write_swa_kv_slice_left_aligned` for prefill
`write_swa_kv_slice_circular` for generate
To clarify, please refer to the picture below. It illustrates scenarios that could exist in our real pipeline.
image

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp (hunk 5 above)
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:149]: Good question. I considered this approach, but ultimately decided not to use it.

Please refer to the picture I attached in the previous comment.

In practice, shifting only happens when the **chunk size is smaller than the window size**. In addition:

1. The chunk size is generally not very small, so the amount of KV data that needs to be moved is relatively large.
2. The number of shifts is very limited and only occurs during the prefill stage, so the impact on overall performance is negligible.
3. Using a circular buffer implementation here would significantly increase the complexity of the logic, making the code harder to maintain in the future. (mask also needs update)

Based on the reasons above, I chose not to use a circular buffer management approach.

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp (hunk 6 above)
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:208]: agree, i have updated code.

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+        // Transposed-V (dim=3) partial-slice shifts degrade into many small remote
+        // transfers. Use a full-buffer round-trip to keep transfer count low:
+        //   1) copy full dst -> CPU tmp,
+        //   2) rearrange entirely on CPU,
+        //   3) copy full CPU tmp -> dst.
+        // This avoids many fine-grained remote transactions in the shift phase.
+        LOG_DEBUG("[SWA] Bulk-shifting KV buffer (dim=3): keeping last " << keep << " of " << old_valid
+                                                                         << " old token(s), capacity=" << capacity);
+        auto whole_tmp =
+            ov::npuw::util::allocMem(dst_tensor->get_element_type(), dst_tensor->get_shape(), "CPU", nullptr);
```
Comment (intelgaoxiong) [inline src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp]: We should have chance to avoid re-allocation.
But it may be not a bottleneck for now.

File: src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp (hunk 6 above)
Comment (intelgaoxiong) [reply src/plugins/intel_npu/src/plugin/npuw/kv_cache_sliding_window_manager.cpp:79]: Thanks. Looks much easier for reading and debugging.

**Review by intelgaoxiong — APPROVED** [review]

**Comment by GuoliangShiIntel** [comment]
Hi @AlexanderKalistratov just a kind reminder that this is the second SWA PR. When you have a chance, could you please take a look? Thank you.

**Review by AlexanderKalistratov — APPROVED** [review]
So it basically assumes that circular buffer is contiguous and not block-based?

**Comment by intelgaoxiong** [comment]
> So it basically assumes that circular buffer is contiguous and not block-based?

You're right. @AlexanderKalistratov 
SWA KVCache has a fixed size, does not increase as seq_len growth.
Block KV management is not needed.
Continuous buffer is enough.
